Empty output while converting `https://graph.microsoft.com/beta/$metadata`

Hello! Thanks for building such a great project!

Currently, I cannot convert the graph API odata metadata to JSON:

```
$ node_modules/odata-openapi/lib/cli.js ./ms-api.xml
```

```json
{"openapi":"3.0.2","info":{"title":"OData CSDL document","description":"","version":""},"paths":{},"components":{"schemas":{}}}
```

`ms-api.openapi3.json (END)`

Thanks for the reply, @ralfhandl !

```
➜  Edge_Downloads npx odata-openapi3 ./metadata.xml
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/odata-openapi3 - Not found
npm ERR! 404
npm ERR! 404  'odata-openapi3@*' is not in this registry.
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in: /Users/[user]/.npm/_logs/2023-06-13T17_38_53_009Z-debug-0.log
➜  Edge_Downloads npm install -g odata-openapi


added 6 packages in 569ms

1 package is looking for funding
  run `npm fund` for details
➜  Edge_Downloads npx odata-openapi3 ./metadata.xml
./metadata.openapi3.json
/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/edm.js:120
        element = element.find(
                          ^

TypeError: Cannot read properties of undefined (reading 'find')
    at module.exports.EDM.#processAnnotations (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/edm.js:120:27)
    at module.exports.EDM.#processSchemas (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/edm.js:101:31)
    at module.exports.EDM.addDocument (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/edm.js:32:25)
    at Object.module.exports.csdl2openapi (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/csdl2openapi.js:69:9)
    at convert (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/cli.js:89:23)
    at Object.<anonymous> (/Users/[user]/.nvm/versions/node/v16.17.1/lib/node_modules/odata-openapi/lib/cli.js:68:3)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
➜  Edge_Downloads
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,11 @@
 # Changelog
 
+## [0.23.1] - 2023-06-15
+
+### Changed
+
+- Annotations for model elements on other schemas are now correctly processed
+
 ## [0.23.0] - 2023-03-31
 
 ### Changed

--- a/lib/csdl2openapi.js
+++ b/lib/csdl2openapi.js
@@ -929,7 +929,7 @@ module.exports.csdl2openapi = function (
         property.typeRefChain.filter((_type) => _type === typeRefChainTail)
           .length > 1
       ) {
-        messages.push(`Cycle detected ${property.typeRefChain.join("->")}`);
+        // messages.push(`Cycle detected ${property.typeRefChain.join("->")}`);
         continue;
       }
 

--- a/lib/edm.js
+++ b/lib/edm.js
@@ -132,12 +132,13 @@ module.exports.EDM = class {
                 (overload.$Parameter?.[0].$Collection
                   ? `Collection(${overload.$Parameter[0].$Type})`
                   : overload.$Parameter[0].$Type || "")) ||
-            (overload.$Parameter || [])
-              .map((p) => {
-                const type = p.$Type || "Edm.String";
-                return p.$Collection ? `Collection(${type})` : type;
-              })
-              .join(",") == args
+            args ==
+              (overload.$Parameter || [])
+                .map((p) => {
+                  const type = p.$Type || "Edm.String";
+                  return p.$Collection ? `Collection(${type})` : type;
+                })
+                .join(",")
         );
       }
       if (!element) {

--- a/lib/edm.js
+++ b/lib/edm.js
@@ -166,9 +166,9 @@ module.exports.EDM = class {
                 Object.assign(parameter, annotations);
               }
             } else {
-              if (element[segments[1]]) {
+              if (typeof element[segments[1]] === "object") {
                 Object.assign(element[segments[1]], annotations);
-              } else {
+              } else if (element.$Kind !== "EnumType") {
                 messages.push(`Invalid annotation target '${target}'`);
               }
             }

--- a/lib/edm.js
+++ b/lib/edm.js
@@ -92,12 +92,17 @@ module.exports.EDM = class {
             });
           }
         } else if (element.$BaseType) {
+          //TODO: will this work if the base type is in a later schema?
           const base = this.namespaceQualifiedName(element.$BaseType);
           if (!this.derivedTypes[base]) this.derivedTypes[base] = [];
           this.derivedTypes[base].push(qualifiedName);
         }
       }
+    }
 
+    // processing annotations requires that all schema elements have been collected in #elements
+    for (const [namespace, schema] of Object.entries(csdl)) {
+      if (!isIdentifier(namespace)) continue;
       this.#processAnnotations(schema.$Annotations ?? {}, messages);
     }
   }
@@ -117,7 +122,7 @@ module.exports.EDM = class {
       } else {
         element = this.element(segments[0].substring(0, open));
         let args = segments[0].substring(open + 1, segments[0].length - 1);
-        element = element.find(
+        element = element?.find(
           (overload) =>
             (overload.$Kind == "Action" &&
               overload.$IsBound != true &&

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "odata-openapi",
-  "version": "0.23.0",
+  "version": "0.23.1",
   "description": "Convert OData CSDL XML or CSDL JSON to OpenAPI",
   "homepage": "https://github.com/oasis-tcs/odata-openapi/blob/master/lib/README.md",
   "bugs": "https://github.com/oasis-tcs/odata-openapi/issues",

--- a/package.json
+++ b/package.json
@@ -22,7 +22,7 @@
   "dependencies": {
     "@discoveryjs/json-ext": "^0.5.7",
     "minimist": "^1.2.8",
-    "odata-csdl": "^0.8.1"
+    "odata-csdl": "^0.8.2"
   },
   "devDependencies": {
     "@apidevtools/openapi-schemas": "^2.1.0",

--- a/test/csdl2openapi.test.js
+++ b/test/csdl2openapi.test.js
@@ -2900,10 +2900,9 @@ describe("Edge cases", function () {
           $UnderlyingType: "Edm.String",
         },
         enumType: {
-          "@Core.LongDescription": "an enumeration type",
           $Kind: "EnumType",
-          foo: {},
-          bar: {},
+          zero: 0,
+          one: 1,
         },
         $Annotations: {
           "typeExamples.single/foo/bar": {
@@ -2914,6 +2913,17 @@ describe("Edge cases", function () {
           },
           "typeExamples.not-there": {
             /* invalid annotation target */
+          },
+          "typeExamples.enumType": {
+            "@Core.LongDescription": "description of enumeration type",
+          },
+          "typeExamples.enumType/zero": {
+            "@Core.LongDescription":
+              "description of enumeration type member has no effect",
+          },
+          "typeExamples.enumType/one": {
+            "@Core.LongDescription":
+              "description of enumeration type member has no effect",
           },
         },
       },
@@ -2963,6 +2973,10 @@ describe("Edge cases", function () {
         },
       },
       "MaxLength"
+    );
+    assert.equal(
+      openapi.components.schemas["typeExamples.enumType"].description,
+      "description of enumeration type"
     );
     assert.deepStrictEqual(
       messages,

--- a/test/csdl2openapi.test.js
+++ b/test/csdl2openapi.test.js
@@ -15,6 +15,8 @@ const assert = require("assert");
 // (external) annotations on actions, functions, parameters, return types
 // control mapping of reference URLs
 
+const { paths, operations, schemas } = require("./utilities");
+
 const { csdl2openapi } = require("odata-openapi");
 
 describe("Edge cases", function () {
@@ -3248,23 +3250,3 @@ describe("Edge cases", function () {
     assert.deepStrictEqual(schemas(actual), schemas(expected), "Schemas");
   });
 });
-
-function paths(openapi) {
-  return Object.keys(openapi.paths).sort();
-}
-
-function operations(openapi) {
-  const p = {};
-  Object.keys(openapi.paths).forEach((template) => {
-    p[template] = Object.keys(openapi.paths[template]).filter(
-      (op) => op != "parameters"
-    );
-  });
-  return p;
-}
-
-function schemas(openapi) {
-  return Object.keys(openapi.components.schemas)
-    .sort()
-    .filter((s) => s.includes("."));
-}

--- a/test/csdl2openapi.test.js
+++ b/test/csdl2openapi.test.js
@@ -495,8 +495,8 @@ describe("Edge cases", function () {
     assert.deepStrictEqual(
       messages,
       [
-        "Cycle detected this.complex1->this.complex2->this.complex1",
-        "Cycle detected this.complex2->this.complex1->this.complex2",
+        // "Cycle detected this.complex1->this.complex2->this.complex1",
+        // "Cycle detected this.complex2->this.complex1->this.complex2",
       ],
       "messages"
     );

--- a/test/funnyInput.test.js
+++ b/test/funnyInput.test.js
@@ -4,9 +4,6 @@ const { paths, operations, schemas } = require("./utilities");
 
 const { csdl2openapi } = require("odata-openapi");
 
-//TODO
-// - annotations targeting enumeration member
-
 describe("Funny input", function () {
   it("Annotation for element in later schema", function () {
     const csdl = {

--- a/test/funnyInput.test.js
+++ b/test/funnyInput.test.js
@@ -1,0 +1,76 @@
+const assert = require("assert");
+
+const { paths, operations, schemas } = require("./utilities");
+
+const { csdl2openapi } = require("odata-openapi");
+
+//TODO
+// - annotations targeting enumeration member
+
+describe("Funny input", function () {
+  it("Annotation for element in later schema", function () {
+    const csdl = {
+      $Version: "4.01",
+      $EntityContainer: "this.Container",
+      this: {
+        $Annotations: {
+          "two.et": {
+            "@Org.OData.Core.V1.Description": "should end up in title",
+          },
+        },
+        Container: {
+          es: {
+            $Type: "other.et",
+            $Collection: true,
+          },
+        },
+      },
+      other: {
+        $Alias: "two",
+        et: {
+          $Kind: "EntityType",
+          $Key: ["key"],
+          key: {},
+          data: {},
+        },
+      },
+    };
+    const expected = {
+      paths: {
+        "/es": {
+          get: {},
+          post: {},
+        },
+        "/es('{key}')": {
+          get: {},
+          patch: {},
+          delete: {},
+        },
+        "/$batch": { post: {} },
+      },
+      components: {
+        schemas: {
+          "other.et": {},
+          "other.et-create": {},
+          "other.et-update": {},
+        },
+      },
+    };
+    const messages = [];
+
+    const actual = csdl2openapi(csdl, { messages });
+    assert.deepStrictEqual(paths(actual), paths(expected), "Paths");
+    assert.deepStrictEqual(
+      operations(actual),
+      operations(expected),
+      "Operations"
+    );
+    assert.deepStrictEqual(schemas(actual), schemas(expected), "Schemas");
+    assert.deepStrictEqual(messages, [], "messages");
+
+    assert.equal(
+      actual.components.schemas["other.et"].title,
+      "should end up in title"
+    );
+  });
+});

